Adding Practice mode

toxix opened the pull request:
This adds an training mode. This should enable you to train with your
database and improve, but do not add your training to the play count. As
the play count should be related to the count that I have played this
song on events/broadcasts.

The design is done in a way that uses the Mixxx's Control System
functionality to comunicate between the gui and the actual counter
implementation.

This is my first contribution to mixxx. So feel free to give me feedback.

Changes should be backwards compatible and original behavior is not changed, when training mode is disabled.

Commit:
Adding Training mode

This adds an training mode. This should enable you to train with your
database and improve, but do not add your training to the play count. As
the play count should be related to the count that I have played this
song on events/broadcasts.

The design is done in a way that uses the Mixxx's Control System
functionality to comunicate between the gui and the actual counter
implementation.

Be-ing commented:
Thanks for working on this. I like the idea. This has bothered me when doing extensive testing during development, especially when playing certain tracks repeatedly in AutoDJ which inflates their play count. I think the UX could use some discussion and refinement. Personally, I would want this on when I open Mixxx, which I do more frequently than actually playing a gig. I always record when I am playing a gig, so it would be convenient for me if this mode automatically turned off when recording starts. I don't use broadcasting, but I presume it would be nice to automatically turn this off when broadcasting too. Thoughts?

I'm not sure about the name "Training Mode", but I don't immediately have any suggestions for a better name.

Holzhaus commented:
Thanks for your contribution!

Do we want to save other stuff like cue point changes in training mode? If not, we could create a copy of the DB when entering training mode and just throw it away when leaving it. I'm torn on this, because you can't use it to prepare hotcue for the next gig. 

Also, I think we should name it "Practice mode" rather than "training mode" IMHO. If we want it on by default, we can just add a "live" button instead (which deactives the implicit practice mode).

In live mode, we could also disable pop-ups that are distracting and can't be closed via a controller (e. g. when you accidently enable vinyl control without a vinyl control input configured).

Be-ing commented:
> Do we want to save other stuff like cue point changes in training mode?

Yes.

> In live mode, we could also disable pop-ups that are distracting and can't be closed via a controller (e. g. when you accidently enable vinyl control without a vinyl control input configured).

These are bad designs and we should get rid of them regardless of this feature.

daschuer commented:
Some time ago we have discussed a "Prepare mode" which streamlines the Gui for the task you do typically at home when preparing the next gig. This could be also interesting.

For this here I am afraid to forget to switch the training mode switch. Without a prominent visual feed back I think we need an undo feature.

The original use case can also be fixed during deleting the history (setlog) playlist.
We may add a "Remove incl. play count" to the context Menu. This might be a good idea in addition anyway.

daschuer commented on src/track/playcounter.h:
```diff
@@ -1,22 +1,19 @@
-    void setTimesPlayed(int iTimesPlayed) {
-        DEBUG_ASSERT(0 <= iTimesPlayed);
-        m_iTimesPlayed = iTimesPlayed;
+  // Sets total number of times a track has been played
```
Please revert this indentation change.
Set you editor to use four space as Tabs.

I also recommend to use git-clang-format to do the code style right. See:
https://www.mixxx.org/wiki/doku.php/coding_guidelines
It just needs some help breaking long lines at colum 80 ... 100.

toxix replied:
This reformate was introduced by git-clang-format. I was also suprised by it. Will remove this, indention change.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -362,6 +362,18 @@ void WMainMenuBar::initialize() {
 #endif
 
+    // TrainingMode
+    m_pConfigTrainingmodeEnabled = new ControlObject(ConfigKey("[TrainingMode]", "enabled"));
```
delete() is missing

toxix replied:
using now std::unique_pointer and std::shared_pointer for that.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -362,6 +362,18 @@ void WMainMenuBar::initialize() {
+    m_pConfigTrainingmodeEnabled = new ControlObject(ConfigKey("[TrainingMode]", "enabled"));
+    QString TrainingmodeTitle = tr("Enable Training Mode");
+    QString TrainingmodeText = tr("In training mode the play counter will not be increased");
+    auto pOptionsTrainingmode = new QAction(TrainingmodeTitle, this);
```
Here you can make use of 
make_parented if you like.

This has just the documentation purpose that the lifetime of this object is managed by the QT object tree.
A explicit delete() would or possible but redundant. It introduces some overhead in object tree management on top.

toxix replied:
Okay, make sense. Will change that also in the rest of this file. Because it's not done with any of the QT objects there, so we have that also consistent in the file.

daschuer commented:
We have now massive log spam during the tests. Don't know why yet.

ronso0 commented:
> The original use case can also be fixed during deleting the history (setlog) playlist.
> We may add a "Remove incl. play count" to the context Menu. This might be a good idea in addition anyway.

Apart from that, a history playlist could have a 'Training' check box. While preparing I may find intersting transitions etc. which I don't want to lose, but disregard the play count when I train a transition multiple times.

ywwg commented:
I agree that this feature is useful but "training" is the wrong word for it.  I would have expected a tutorial mode that walks the use through how mixxx works (also useful!).  "Prepare" is good as Daniel suggested.  Maybe "Practice Mode"? "Rehearsal"?

daschuer commented:
>Apart from that, a history playlist could have a 'Training' check box. While preparing I may find intersting transitions etc. which I don't want to lose, but disregard the play count when I train a transition multiple times.

Interesting how different we think about the mode switch. I am afraid this needs to be part of a whole concept. Since this is a lot of work. How about solve the original use case by special history remove features:

"Remove, keep playcount" (original, is there a valid use case?
"Remove, reset playcount"
"Reset playcount"
"Reset repeat playcount"

For the reset playcount feature we can introduce a new playlist mode. "History without playcount"

This has the benefit that the user can decide at any time how to store history and playcount.

What do you think?

Commit:
Revert indention change

Revert the indention changes introduced in
edd1525db7f2e477a2993d6b5ae0b8d3f5762b9a
We should use 4 spaces as indention as stated in comment from
https://github.com/mixxxdj/mixxx/pull/2431/files/edd1525db7f2e477a2993d6b5ae0b8d3f5762b9a#r364600409
by daschuer.

Commit:
Change to use make_parented

As stated by dashauer in the codereview it make sense to
use the make_parented as this indicates that the live time
of the object it controlled by the qt parrent.

reference:
https://github.com/mixxxdj/mixxx/pull/2431/files/edd1525db7f2e477a2993d6b5ae0b8d3f5762b9a#r364600990

Commit:
Fix possible memory leaks

In edd1525db7f2e477a2993d6b5ae0b8d3f5762b9a there have been
introduced some possible memory leakes. This should be fixed
with using of std::unique_pointers and std::shared_pointers.

reference:
https://www.mixxx.org/wiki/doku.php/coding_guidelines#pointer_object_lifetime_ownership

ronso0 commented:
> I would have expected a tutorial mode that walks the use through how mixxx works (also useful!).

Thought about the same when I read the PR title...

> "Remove, keep playcount" (original, is there a valid use case?
"Remove, reset playcount"
"Reset playcount"
"Reset repeat playcount"

Yes, this a full list IMO. No use case for the origial implementation ("Remove, keep playcount") comes to mind, we should drop it.

Yet another idea how to interact with the feature:
* any new history playlist is created as 'training' list with a grey icon (Play or just **`*`**)
* list mode can be switched later on to Live, then play count is added to track info

Holzhaus commented:
Maybe we can simplify the whole thing. We could use a temporary history playlist by default. When closing Mixxx it will ask you "do you want to save history?".

I think that's intuitive and reduces the risk of losing/saving history because you're in the wrong mode. 

Additionally, we could also right click the temporary history playlist while mixxx is running and have a "Save" context menu item, so that the current state of the temporary history is applied.

daschuer commented:
We already have a pop up when closing Mixxx. I am not in favour to introduce a second one. 
I think the intermediate history state of rons0 will work. Just check the played state during a session. And count plays via a right click. Or the other way round to not suprise legacy users. 
This would require a new history playlist type, one with and one without counts. And then we can go for and back at any time.

This would allow also to reset the cound of playlist that are very old without loosing them.

poelzi commented:
I thought about this feature now for a while and would suggest following change:
Instead of changing the playcount behaviour, we should simply add a new counter called livecount that only increases when played in live mode. I think this feature in general is useful, but fideling with the play counter is only confusing and problematic.
I for example build my sets always on site to better catch the vibe of the atmosphere. I need the play counter to stay this way because I try not to play tracks with 0 plays because they might suck. I only try to mix harmonic so I often end up with very complex filters and only a handful of tracks could possible fit, not all have stars.
Having a separate counter allows you to nicely filter out often played tracks or we could then even implement a special filter for filter out tracks that where played live under a certain age.

Commit:
Rename 'training mode' to 'practice mode'

The name training mode is more accosiated with an introduction
into mixxx. But this is not what this feature is doing. So to
get the expectations clear. I follow the suggestion of the PR
and rename this feature to practice mode.

reference:
https://github.com/mixxxdj/mixxx/pull/2431#issuecomment-572805476

toxix commented:
Hey Guys,
Looks like there are more thoughts on this and how to continue this further. For now I have make the code free from memory leaks and renamed the feature to 'practice mode'. Thanks for the suggestion and input on the confusing naming.

Additionaly I have rebased the branch onto current master so there are no merge conflicts anymore.

Suggestion is to merge this for now and I think here are a lot of good ideas so feel free to take it over and improve it further. Like for example using the playlist history instead of the static playcounter and add tagging to the playlist so that you can dynamicaly filter for tracs on that. I might actualy come back on this. As this doing the job for me at this moment and it was a small improvement to learn about the codebase.

I am currently trying to improve the way of how tagging/creates work and than improve searching/filtering for the library. Where do you discuss concepts? So I can collect some ideas early on.

Commit:
Rename config key

Rename move the practiceMode config key scope into the library. As
its a better fit there, than create a new config key group for this
single setting.

toxix commented on src/track/playcounter.cpp:
```diff
@@ -1,7 +1,20 @@
+PlayCounter::PlayCounter(int timesPlayed)
+        : m_iTimesPlayed(timesPlayed),
+          m_bPlayed(false) {
+    m_pTrainingmodeEnabled = new ControlProxy("[TrainingMode]", "enabled");
```
Using unique pointers and shared pointers as suggested. Also using the [Library] group for the single setting now.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -109,11 +109,11 @@ void WMainMenuBar::initialize() {
 
     // LIBRARY MENU
-    QMenu* pLibraryMenu = new QMenu(tr("&Library"));
+    QMenu* pLibraryMenu = make_parented<QMenu>(tr("&Library"));
```
auto

toxix replied:
updated

daschuer commented on src/track/playcounter.cpp:
```diff
@@ -1,7 +1,20 @@
+PlayCounter::PlayCounter(int timesPlayed)
+        : m_iTimesPlayed(timesPlayed),
+          m_bPlayed(false) {
+    m_pPracticemodeEnabled = std::make_shared<ControlProxy>("[Library]", "practice_mode");
```
Is this pointer actually shared? 
Could we use a unique_ptr?

toxix replied:
yes it is, The playCounter is copied/assigned at other points in the codebase, you will get an compile error, if you are using a unique_ptr at this place.

daschuer commented:
Than you for working on this. 

Before we can merge your work, we need your permission. Please sign: https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ

... and comment here when done.

Normally we do not merge features until they are finished. In this case I am unsure. 
While this works and is already an improvement, I am in favour of a solution, that can be used at any time. What do the other think? Merge?

In general we discuss development concepts on Zulip. Ideas are collected as bugs at https://bugs.launchpad.net/bugs

As of the library, there is a development in progress switching to aoide, a new library concept. Any help is welcome.

ronso0 commented:
> Normally we do not merge features until they are finished. In this case I am unsure.
While this works and is already an improvement, I am in favour of a solution, that can be used at any time. What do the other think? Merge?

I don't think we should merge this.
There are a few objections about the implementation proposed here, and even more ideas how to suit the individual needs. Maybe all desired options should be available in the Library prefernces?
* play count vs. marker for 'played once, Cue set' vs. Live play count (played while broadcasting or recording?)
* default on vs. default Off
* ...

and use cases that are better served with context menu actions:
* delete playlist, different options to handle play count
* split current log now
* switch log mode: prep vs. Live, maybe in combination with split?
* ...

Commit:
Refector for codreview

using auf 'auto' keyword
simplify if statement

toxix commented:
@daschuer Update the PR and signed the thing

@ronso0 Feel free to extend this code with the things you see as required. I see your point of having even more features extend the use cases. But as soon as things are usable and working they should be used in my opinion. There is nothing that is stopping you from extending this feature, no incompatibilety to an extended feature, so do you want to take ownership of this extended features you mentioned?

ronso0 commented:
This is not an urgent bugfix so I don't see a reason to hurry merging this a few days after opening the PR just becasue it works, considering the the opinions shared above.

I agree with @ywwg that Practice mode fits better.

We should at least take the time to add a visual indicator to the GUI. Making the Practice Mode toggle a single menubar item with a checkbox could work IMO, but idk if that's desired and works on all OS. Alternatively, altering the 'Current history' icon is a no-brainer, the child item decoration happens [here](https://github.com/mixxxdj/mixxx/blob/master/src/library/setlogfeature.cpp#L192)
A quick solution would be to replace the curret arrow **>** with a
**▷** (grey) icon for Practice mode
**⯈** (orange) icon for Live mode

The delete options can be postponed IMO.

The actions to split off / re-merge the current history when toggling the Practice/Prep mode would be nice-to-have as it avoids having histories with mixed play counts. The actions to accomplish this are already there: [`Create new history playlist`](https://github.com/mixxxdj/mixxx/blob/master/src/library/setlogfeature.cpp#L47) and [`Join with previous`](https://github.com/mixxxdj/mixxx/blob/master/src/library/setlogfeature.cpp#L41) in [SetlogFeature::onRightClickChild](https://github.com/mixxxdj/mixxx/blob/master/src/library/setlogfeature.cpp#L103)

What do others think?

daschuer commented:
Sounds good. I just prefer to introduce new playlist (hidden type) for the Practice mode. 
This way we can distinguish them later and keep the options to build features on it.

ywwg commented on src/track/playcounter.cpp:
```diff
@@ -1,7 +1,20 @@
+    // if we are in PracticeMode mode, the play count should not be changed
+    // so we can return here
+    bool practicemode_enabled = m_pPracticemodeEnabled->toBool();
+    if (practicemode_enabled) {
```
it's fine to just say if(m_pPracticemodeEnabled->toBool())

ywwg commented on src/widget/wmainmenubar.cpp:
```diff
@@ -357,13 +357,25 @@ void WMainMenuBar::initialize() {
 #endif
 
+    // PracticeMode
+    m_pConfigPracticemodeEnabled = std::make_unique<ControlObject>(ConfigKey("[Library]", "practice_mode"));
```
this initialization should be done in the constructor

ywwg commented on src/track/playcounter.h:
```diff
@@ -40,6 +39,7 @@ class PlayCounter {
 private:
     int m_iTimesPlayed;
     bool m_bPlayed;
+    std::shared_ptr<ControlProxy> m_pPracticemodeEnabled;
```
The style in mixxx is just to use ControlProxy*.  Don't worry about freeing the pointer we handle that.

toxix replied:
How is that handeled? like when I look into that comment it's like hey take care of your things: https://github.com/mixxxdj/mixxx/pull/2431#discussion_r364600741

ywwg replied:
oh oops.  hmm.

ywwg replied:
I don't know what our policy is on using sharedpointer to avoid needing to add things to the destructor.  @uklotzde ? @daschuer ?

ywwg commented on src/widget/wmainmenubar.h:
```diff
@@ -78,13 +81,15 @@ class WMainMenuBar : public QMenuBar {
 
     UserSettingsPointer m_pConfig;
     ConfigObject<ConfigValueKbd>* m_pKbdConfig;
+    std::unique_ptr<ControlObject> m_pConfigPracticemodeEnabled;
```
just make this a member:
`ControlObject m_pConfigPracticemodeEnabled;`

no need for it to be a pointer

ywwg replied:
this comment is also wrong, sorry.  We usually make controlobjects pointers too (and then have to delete them).  Again, not sure why we haven't started migrating toward unique/shared pointers but it hasn't been the style yet.  Do we want to mix things?

ywwg reviewed (changes requested):
thanks for working on this, here are some small style changes

poelzi commented:
Actually, I think inverting the concept would actually be much better and introduce the "live" mode.
I don't know how you use mixxx, but for me it's my sole music player and basically always on. At some points, for very few hours, I'm in public/broadcast mode, aka live.

For live mode, I would like to see  behavior like this:
- Ensure a new playlist is started when enter/exit live mode.
- I want (maybe depending on setting), that the recording starts as soon as I starting a deck. No more juggling start recording and starting track etc at the same time
- I mark the history item as such, different icon would me nice
- Protect the history item by default
- live_played counter that counts the number of live performances.
- Maybe some virtual value for last_live_played for queries like last_live_played > 6 months.
- Maybe a filter on the history browser to only show live histories

ronso0 commented:
I'll close this abandoned PR now to clean up the list of open PRs.
If the author or mainatiners think this should remain open please reopen.
If anyone thinks this is worth salvaging and is motivated to pick it up, resolve the conflicts and move it forward, feel free to open a new PR.

ronso0 commented:
Related PR potentially superseding this: #10394

ronso0 commented:
Or my more granular approach in #16170
